src/SonarScanner.MSBuild.Common/ProcessRunnerArguments.cs

```diff
@@ -216,17 +216,22 @@ private static string ShellEscape(string argLine)
             var sb = new StringBuilder();
             foreach (var c in argLine)
             {
-                // This escape is required after %* is expanded to prevent command injections
-                sb.Append('^');
-                sb.Append('^');
-
-                // This escape is required only to pass the argument line to the batch script
-                sb.Append('^');
+                if (IsSpecialCharacter(c))
+                {
+                    sb.Append('^');
+                }
                 sb.Append(c);
             }
             return sb.ToString();
         }
+        private static bool IsSpecialCharacter(char c)
+        {
+            // Define the special characters that need escaping
+            char[] specialChars = { '^', '>', '<', '|', '&' };
```

Comment:
Can you please create a hashset of special characters as a field in the class and check against that?

Reply:
Done.